##PRAGYA RAGHUVANSHI
Net ID: pr158

###Installing dependencies and packages

In [21]:
%%capture
!apt-get install ffmpeg freeglut3-dev xvfb  # For visualization
!pip install pyglet==1.4
!pip install tensorflow==2.0.0-alpha0
!pip install autorom
!AutoROM --accept-license


In [22]:
%%capture
!git clone https://github.com/DLR-RM/stable-baselines3 && cd stable-baselines3
!pip install -e .[docs,tests,extra]

In [23]:
## Library Imports
import gym
import numpy as np
from stable_baselines3 import A2C
from stable_baselines3.common.env_util import make_vec_env
from stable_baselines3.common.evaluation import evaluate_policy
import tensorflow as tf
import warnings
warnings.filterwarnings('ignore')
%load_ext tensorboard


logdir = ('./cartpole_tb_3')
tensorboard_callback = tf.keras.callbacks.TensorBoard(logdir, histogram_freq=1)

The tensorboard extension is already loaded. To reload it, use:
  %reload_ext tensorboard


###GYM Environments For Training & Evaluation

In [24]:
#HYPERPARAMETER
training_steps = 10000

In [25]:
## Training environment of Cartpole
train_env = gym.make('CartPole-v1')

## Evalutation environment of Cartpole
evaluation_env = gym.make('CartPole-v1')

In [26]:
#vanilla model
van_model = A2C('MlpPolicy', train_env, verbose = 0, tensorboard_log=logdir , gae_lambda = 0) 

# GAE advantage model
gae_model = A2C('MlpPolicy', train_env, verbose = 0, tensorboard_log=logdir , gae_lambda = 0.85)

#n-step advantage model
nstep_model = A2C('MlpPolicy', train_env, verbose = 0, tensorboard_log=logdir , gae_lambda = 1, n_steps = 5)

### Evaluating Rewards of the model 
1. Vanilla Model

In [27]:
# training 
van_model.learn(total_timesteps=training_steps)
#evaluating performance
vanillareward_mean, vanillareward_std= evaluate_policy(van_model, evaluation_env, n_eval_episodes=100)



2. GAE model 

In [28]:
# training 
gae_model.learn(total_timesteps=training_steps)
#evaluating performance
gaereward_mean, gaereward_std= evaluate_policy(gae_model, evaluation_env, n_eval_episodes=100)


3. n-STEP MODEL

In [29]:
# training 
nstep_model.learn(total_timesteps=training_steps)
#evaluating performance
nstepreward_mean, nstepreward_std= evaluate_policy(nstep_model, evaluation_env, n_eval_episodes=100)


###Model Results

In [30]:
print(f'Mean Reward of Vanilla model: {np.round(vanillareward_mean,2)} +/- {np.round(vanillareward_std,2)}')
print(f'Mean Reward of N-Step model: {np.round(nstepreward_mean,2)} +/- {np.round(nstepreward_std,2)}')
print(f'Mean Reward of GAE model: {np.round(gaereward_mean,2)} +/- {np.round(gaereward_std,2)}')

Mean Reward of Vanilla model: 52.31 +/- 22.15
Mean Reward of N-Step model: 379.06 +/- 63.69
Mean Reward of GAE model: 500.0 +/- 0.0


###Analysis
From the analysis we can see that the GAE model has the best performance with the mean reward of 500. The N-Step is the second best model and fits with a mean reward of 379.06 Also, the Vanilla model performs the worst with a mean reward of 52.31.  These values will likely vary as the cells are rerun. 
